test/system/505-networking-pasta.bats

```diff
@@ -209,34 +209,46 @@ function pasta_test_do() {
         local expect="$(for i in $(seq ${seq}); do printf "x"; done)"
     fi
 
-    # Set retry/initial delay for client to connect
-    local delay=1
-    if [ ${ip_ver} -eq 6 ] && [ "${addr}" != "::1" ]; then
-        # Duplicate Address Detection on link-local
-        delay=3
-    elif [ "${proto}" = "udp" ]; then
-        # With Podman up, and socat still starting, UDP clients send to nowhere
-        delay=2
-    fi
+    # start server
+    local cname="c-socat-$(safename)"
+    run_podman run -d --name="$cname" --net=pasta"${pasta_spec}" -p "${podman_spec}" "${IMAGE}" \
+                   sh -c 'for port in $(seq '"${xseq}"'); do '\
+'                             socat -u '"${bind}"' '"${recv}"' & '\
+'                         done; wait'
 
-    # Now actually run the test: client,
-    for one_port in $(seq ${seq}); do
-        local connect="${proto_upper}${ip_ver}:[${addr}]:${one_port}"
-        [ "${proto}" = "udp" ] && connect="${connect},shut-null"
+    run_podman container inspect --format "{{.NetworkSettings.SandboxKey}}" $cname
+    netns="$output"
+
+    # Make sure all ports in the contianer are bound
+    for cport in $(seq ${xseq}); do
+        retries=50
+        while [[ $retries -gt 0 ]]; do
+            run_podman unshare nsenter -n"$netns" ss -Htuln -$ip_ver sport = $cport
```

Comment:
True, but this is on the "fast path", and `ss` isn't always cheap, because it has to fetch the whole procfs entry:

```
$ unshare -rUn
# time for i in $(seq 1 1000); do ss -Htuln -4 sport = 80 >/dev/null; done

real	0m10.573s
user	0m0.248s
sys	0m10.356s
# time for i in $(seq 1 1000); do ss -Huln -4 sport = 80 >/dev/null; done

real	0m1.178s
user	0m0.869s
sys	0m0.343s
# time for i in $(seq 1 1000); do ss -Htln -4 sport = 80 >/dev/null; done

real	0m10.576s
user	0m0.318s
sys	0m10.289s
```

And it's not negligible either for 50 retries:
```
# time for i in $(seq 1 50); do ss -Htln -4 sport = 80 >/dev/null; done

real	0m0.522s
user	0m0.022s
sys	0m0.502s
# time for i in $(seq 1 50); do ss -Huln -4 sport = 80 >/dev/null; done

real	0m0.073s
user	0m0.048s
sys	0m0.027s
```

Reply:
ss doesn't fetch anything from procfs? It uses netlink (NETLINK_SOCK_DIAG) to get the sockets.

Ok I agree if there are timing benefits, actually no need for special parsing in the test I can add --$proto instead of using the short option.

Reply:
Aah, right, it uses `NETLINK_SOCK_DIAG` for everything as long as `tcp_show_netlink()` succeeds. Still, the cost doesn't come from the interface, but rather from the kernel preparing the TCP socket table dump (UDP is somewhat more straightforward).